string MAC
string pickled_message